tf_quant_finance/black_scholes/vanilla_prices.py

```diff
@@ -182,6 +182,262 @@ def option_price(*,
+    is_call_options: A real `Tensor` of `boolean` values and of the shape
+      that broadcasts with `volatilities`. True if option is call else
+      false.
+    dtype: Optional `tf.DType`. If supplied, the dtype to be used for conversion
+      of any supplied non-`Tensor` arguments to `Tensor`.
+      Default value: `None` which maps to the default dtype inferred by
+      TensorFlow.
+    name: str. The name for the ops created by this function.
+      Default value: `None` which is mapped to the default name `option_price`.
+  Returns:
+    option_prices: A `Tensor` of same shape as `spots`. The approximate price of
+    the barriers option under black scholes.
+
+  """
+  if (continuous_dividends is None) == (cost_of_carries is None):
+    raise ValueError('At most one of continuous_dividends and cost of carries '
+                     'may be supplied')
+  with tf.name_scope(name or "barrier_price"):
+    # Convert all to tensor and enforce float dtype where required
+    if discount_rates is not None:
+      discount_rates = tf.convert_to_tensor(
+          discount_rates, dtype=dtype, name="discount_rates")
+    else:
+      discount_rates = tf.convert_to_tensor(
+          1, dtype=dtype, name="discount_rates")
+
+    dtype = discount_rates.dtype
+
+    if continuous_dividends is not None:
+      continuous_dividends = tf.convert_to_tensor(
+          continuous_dividends, dtype=dtype, name="continuous_dividends")
+
+    if cost_of_carries is not None:
+      continuous_dividends = tf.convert_to_tensor(
+          discount_rates-cost_of_carries, dtype=dtype,
+          name="continuous_dividends")
+
+    spots = tf.convert_to_tensor(
+        spots, dtype=dtype, name="spots")
+    strikes = tf.convert_to_tensor(
+        strikes, dtype=dtype, name="strikes")
+    barriers = tf.convert_to_tensor(
+        barriers, dtype=dtype, name="barriers")
+    rebates = tf.convert_to_tensor(
+        rebates, dtype=dtype, name="rebates")
+    volatilities = tf.convert_to_tensor(
+        volatilities, dtype=dtype, name="volatilities")
+    expiries = tf.convert_to_tensor(
+        expiries, dtype=dtype, name="expiries")
+
+    is_barrier_down = tf.convert_to_tensor(is_barrier_down, dtype=tf.int32,
+                                           name="is_barrier_down")
+    is_knock_out = tf.convert_to_tensor(is_knock_out, dtype=tf.int32,
+                                        name="is_knock_out")
+    is_call_options = tf.convert_to_tensor(is_call_options, dtype=tf.int32,
+                                           name="is_call_options")
+    indices = tf.bitwise.left_shift(
+        is_barrier_down, 2) + tf.bitwise.left_shift(
+            is_knock_out, 1) + is_call_options
+
+    # Masks select the appropriate terms for integral approximations
+    # Interals are seperated by algebraic terms and probability
+    # distribution terms. This give 12 different terms per matrix
+    # (6 integrals, 2 terms each)
+    # shape = [1, 12]
```

Comment:
Should this be [8, 12]?

Reply:
I was referencing for each row, but you're right should be [8, 12] for matrix. Modified it!